src/Gaffer/ValuePlug.cpp

```diff
@@ -244,83 +217,114 @@ class ValuePlug::HashProcess : public Process
-				if( threadData.clearCache.load( std::memory_order_acquire ) )
+			const bool forceMonitoring = Process::forceMonitoring( threadState, plug, staticType );
+
+			auto acquireHash = [&]( const HashCacheKey &cacheKey ) {
+
+				// Before we use this key, check that the dirty count hasn't maxed out
+				if(
+					cacheKey.dirtyCount == DIRTY_COUNT_RANGE_MAX ||
+					cacheKey.dirtyCount == DIRTY_COUNT_RANGE_MAX + 1 + DIRTY_COUNT_RANGE_MAX
+				)
 				{
-					threadData.cache.clear();
-					threadData.clearCache.store( 0, std::memory_order_release );
+					throw IECore::Exception(  "Dirty count exceeded max. Either you've left Gaffer running for 100 million years, or a strange bug is incrementing dirty counts way too fast." );
 				}
 
-				if( threadData.cache.getMaxCost() != g_cacheSizeLimit )
+				// Check for an already-cached value in our thread-local cache, and return it if we have one.
+				if( !forceMonitoring )
 				{
-					threadData.cache.setMaxCost( g_cacheSizeLimit );
+					if( auto result = threadData.cache.getIfCached( cacheKey ) )
+					{
+						return *result;
+					}
 				}
 
-				// And then look up the result in our cache.
-				if( g_hashCacheMode == HashCacheMode::Standard )
+				// No value in local cache, so either compute it directly or get it via
+				// the global cache if it's expensive enough to warrant collaboration.
+				IECore::MurmurHash result;
+				if( cachePolicy == CachePolicy::Legacy || cachePolicy == CachePolicy::Standard )
 				{
-					return threadData.cache.get( processKey, currentContext->canceller() );
+					result = HashProcess( p, plug, computeNode ).run();
 				}
-				else if( g_hashCacheMode == HashCacheMode::Checked )
+				else
 				{
-					HashProcessKey legacyProcessKey( processKey );
-					legacyProcessKey.dirtyCount = g_legacyGlobalDirtyCount + DIRTY_COUNT_RANGE_MAX + 1;
-
-					const IECore::MurmurHash check = threadData.cache.get( legacyProcessKey, currentContext->canceller() );
-					const IECore::MurmurHash result = threadData.cache.get( processKey, currentContext->canceller() );
-
-					if( result != check )
+					std::optional<IECore::MurmurHash> cachedValue;
+					if( !forceMonitoring )
+					{
+						cachedValue = g_cache.getIfCached( cacheKey );
+					}
+					if( cachedValue )
 					{
-						// This isn't exactly a  process exception, but we want to treat it the same, in
-						// terms of associating it with a plug.  Creating a ProcessException is the simplest
-						// approach, which can be done by throwing and then immediately wrapping
-						try
-						{
-							throw IECore::Exception(  "Detected undeclared dependency. Fix DependencyNode::affects() implementation." );
-						}
-						catch( ... )
-						{
-							ProcessException::wrapCurrentException( processKey.plug, currentContext, staticType );
-						}
+						result = *cachedValue;
+					}
+					else
+					{
+						result = Process::acquireCollaborativeResult<HashProcess>( cacheKey, p, plug, computeNode );
```

Comment:
Yeah, this does feel a bit fiddly to me too. I don't think your suggestion quite works as-is though because we don't want to `return` but rather assign `result = `. So we'd require another level of nesting with the cache lookup in an `else` block, which would reduce its clarity a bit.

I did do what I did fairly consciously because I noticed a pre-existing [bug](https://github.com/GafferHQ/gaffer/blob/main/src/Gaffer/ValuePlug.cpp#L647-L649) where we were failing to perform any kind of TaskCollaboration when forcing monitoring. So I felt the slightly more awkward construction was worth it to preclude the possibility of that sort of thing happening again.

So on balance, I'd like to keep this as it is.

Reply:
> I do appreciate more you wanting to just rip out the forceMonitoring stuff now

It's not the `forceMonitoring` stuff that I want to rip out - that's essential. What I'm fairly keen to rip out is the `HashCacheMode` stuff, but I thought that should at least wait until a new major version.

Reply:
Good points. It makes that this logic is actually a bit complex - it probably should actual give a bit more pause when reading it than most 2 if blocks.